core-bundle/src/Controller/FrontendModule/TwoFactorController.php

```diff
@@ -0,0 +1,181 @@
+
+        $redirectPage = $adapter->findByPk($model->jumpTo);
+        $return = $redirectPage instanceof PageModel ? $redirectPage->getAbsoluteUrl() : $this->page->getAbsoluteUrl();
+
+        $template->error = false;
+        $template->action = '';
+        $template->enforceTwoFactor = $this->page->enforceTwoFactor;
+        $template->targetPath = $return;
+
+        // Inform the user if 2FA is enforced
+        if ($this->page->enforceTwoFactor) {
+            $template->message = $this->translator->trans('MSC.twoFactorEnforced', [], 'contao_default');
+        }
+
+        if ((!$user->useTwoFactor && $this->page->enforceTwoFactor) || 'enable' === $request->get('2fa')) {
+            $this->enableTwoFactor($template, $request, $user, $return);
+        }
+
+        if (!$this->page->enforceTwoFactor && 'tl_two_factor_disable' === $request->request->get('FORM_SUBMIT')) {
+            $this->disableTwoFactor($user);
+        }
+
+        $template->isEnabled = $user->useTwoFactor;
+        $template->href = $this->router->generate('tl_page.'.$this->page->id, ['2fa' => 'enable']);
+        $template->twoFactor = $this->translator->trans('MSC.twoFactorAuthentication', [], 'contao_default');
+        $template->explain = $this->translator->trans('MSC.twoFactorExplain', [], 'contao_default');
+        $template->active = $this->translator->trans('MSC.twoFactorActive', [], 'contao_default');
+        $template->enableButton = $this->translator->trans('MSC.enable', [], 'contao_default');
+        $template->disableButton = $this->translator->trans('MSC.disable', [], 'contao_default');
+
+        return $template->getResponse();
+    }
+
+    private function enableTwoFactor(Template $template, Request $request, FrontendUser $user, $return): void
+    {
+        // Return if 2FA is enabled already
+        if ($user->useTwoFactor) {
+            return;
+        }
+
+        /** @var AuthenticationException|null $exception */
+        $exception = $this->authenticationUtils->getLastAuthenticationError();
+
+        if ($exception instanceof InvalidTwoFactorCodeException) {
+            $template->error = true;
+            $template->message = $this->translator->trans('ERR.invalidTwoFactor', [], 'contao_default');
+        }
+
+        // Validate the verification code
+        if ('tl_two_factor' === $request->request->get('FORM_SUBMIT')) {
+            if ($this->authenticator->validateCode($user, $request->request->get('verify'))) {
+                // Enable 2FA
+                $user->useTwoFactor = '1';
+                $user->save();
+
+                throw new RedirectResponseException($return);
+            }
+
+            $template->error = true;
+            $template->message = $this->translator->trans('ERR.invalidTwoFactor', [], 'contao_default');
+        }
+
+        // Generate the secret
+        if (!$user->secret) {
+            $user->secret = random_bytes(128);
```

Comment:
Shouldn't we generate a new secret whenever 2FA is enabled, and not when the module is shown?

Reply:
It will be generated if 2FA gets enabled (or enforced).